proposals/0387-cross-compilation-destinations.md

```diff
@@ -80,85 +86,93 @@ The proposed structure of artifact bundles containing CC destinations looks like
 
-Note the presence of `destination.json` files in each `<host variant>` subdirectory. These files should contain a JSON
-dictionary with an evolved version of the schema of [existing `destination.json` files that SwiftPM already
-supports](https://github.com/apple/swift-package-manager/pull/1098) and `destination.json` files presented in the pitch
-version of this proposal, hence `"schemaVersion": "3.0"`:
+Here's how `info.json` file could look like for `swift-5.8_ubuntu.artifactbundle` introduced in the example
+above:
 
-```json
+```json5
 {
-  "schemaVersion": "3.0",
-  "buildTimeTriples": ["<an array of supported build-time platform triples>"],
-  "runTimeTriples": ["<an array of supported run-time platform triples>"],
-  "swiftResourcesPaths": ["<array with relative paths containing Swift resources in the destination tree>"],
-  "includeSearchPaths": ["<array with relative paths containing headers in the destination tree>"],
-  "librarySearchPaths": ["<array with relative paths containing libraries in the destination tree>"],
+  "artifacts" : {
+    "swift-5.8_ubuntu22.04" : {
+      "type" : "crossCompilationDestination",
+      "version" : "0.0.1",
+      "toolset": "toolset.json",
+      "variants" : [
+        {
+          "path" : "ubuntu_jammy",
+          "supportedTriples" : [
+            "arm64-apple-darwin",
+            "x86_64-apple-darwin"
+          ]
+        }
+      ]
+    },
+    "swift-5.8_ubuntu20.04" : {
+      "type" : "crossCompilationDestination",
+      "version" : "0.0.1",
+      "variants" : [
+        {
+          "path" : "ubuntu_focal",
+          "supportedTriples" : [
+            "arm64-apple-darwin",
+            "x86_64-apple-darwin"
+          ]
+        }
+      ]
+    }
+  },
+  "schemaVersion" : "1.0"
 }
 ```
 
-Thanks to the availability of Universal Binaries on macOS and multiarch layouts on Linux, `buildTimeTriples` and
-`runTimeTriples` properties use plural in their naming and their values are arrays.
-
-We propose that all relative paths in `destination.json` files should be validated not to "escape" the destination
-bundle for security reasons. That is, `../` components, if present in paths, will not be allowed to reference files and
-directories outside of a corresponding destination bundle. Symlinks will also be validated to prevent them from escaping
-out of the bundle.
-
-Since not all platforms can support self-contained destination bundles, users will be able to provide their own
-additional paths on the filesystem outside of bundles after a destination is installed. The exact options for specifying
-paths are proposed in a subsequent section for a newly introduced `swift destination configure` command.
-
-### `toolset.json` files
+### `toolset.json` Files
 
 We find that properties dedicated to tools configuration are useful outside of the cross-compilation context. Due to
-that, a separate `toolset.json` file is introduced:
+that, separate toolset configuration files are introduced:
```

Comment:
Now that you're removing the host variant directories, have you thought about how to have these bundles use different tools based on the host, particularly on linux? Would be good to specify that, whether you plan to have different toolsets based on the build-time triple or place them in certain directories by convention.

Reply:
OK, I took a look at that SE-0305 proposal, that I was not familiar with. It specifically mentions listing out paths to different tools based on the build-time triple needed, which is missing from this proposal. That creates an ambiguity after this pull, where you don't know which build-time triple each of these toolsets is built for, that wasn't there with the host variant directories before, with each presumably containing tools that run on that build host.

Reply:
Yes, it does that, but then check out example code for SwiftPM command plugins that utilize artifact bundles. Host variabt directories are not present there and looks like it's convention-based after all.

Reply:
Sorry, not command plugins, but extensible build tools. I'm referring to this one https://github.com/abertelrud/swiftpm-buildtool-plugin-examples/tree/main/Example%201%20-%20SwiftGen/Binaries/SwiftGenBinary.artifactbundle

Reply:
That's okay for that example because in that context `supportedTriples` refers to runtime triples, which are all presumably supported by that single binary. Here, the `supportedTriples` refers to build-time triples, so you need separate executables for each build host, as in the protocol buffer example I mentioned.

Before this pull, you had a `toolset.json` in each host variant directory, so SPM could install the right toolset for the build host. Now, there's no way to differentiate the tools for different build-time triples, so this pull will not be enough.

I misspoke about "convention" before: there has to be an explicit way for bundlers to communicate to SPM where the tools are located for each build-time triple, which is no longer there after this pull.

Reply:
> I misspoke about "convention" before: there has to be an explicit way for bundlers to communicate to SPM where the tools are located for each build-time triple, which is no longer there after this pull.

There is, for build-time platforms with universal binaries put all build-time triples supported by such binaries into `supportedTriples` array for a single variant:

```
{
  "artifacts" : {
    "swift-5.8_ubuntu22.04" : {
      "type" : "crossCompilationDestination",
      "version" : "0.0.1",
      "variants" : [
        {
          "path" : "ubuntu_jammy",
          "supportedTriples" : [
            "arm64-apple-darwin",
            "x86_64-apple-darwin"
          ]
        }
      ]
    }
  }
}
```

For other build-time platforms just create a different variant with a different path with a single element in the `supportedTriples` array:

```
{
  "artifacts" : {
    "swift-5.8_ubuntu22.04" : {
      "type" : "crossCompilationDestination",
      "version" : "0.0.1",
      "variants" : [
        {
          "path" : "ubuntu_jammy-bt-aarch64",
          "supportedTriples" : ["aarch64-unknown-linux-gnu"]
        },
        {
          "path" : "ubuntu_jammy-bt-x86_64",
          "supportedTriples" : ["x86_64-unknown-linux-gnu"]
          ]
        }
      ]
    }
  }
}
```

Reply:
I think it would be better to add a `buildTimePath` field to the toolset.json spec, to tell SPM to look in a specific subdirectory of `toolsetRootPath` for the tools for a particular build-time triple. That way we wouldn't have to create more destinations for linux like this and you could use those toolsets without bundles too.

Reply:
IMO that's over-engineering toolsets for what they weren't designed for (a SwiftPM's answer to CMake toolchains, and those don't support multiple triples in a single toolchain file either). If you need different tools for a different build-time or run-time triple and Universal binaries aren't available, just create a separate toolset.

> and you could use those toolsets without bundles too.

I don't think there's anything that prevents using toolsets without bundles in their current state.

Reply:
> If you need different tools for a different build-time or run-time triple and Universal binaries aren't available, just create a separate toolset.

As you said above, that won't be enough for bundles, a whole new destination is needed.

> I don't think there's anything that prevents using toolsets without bundles in their current state.

I said "those toolsets," meaning a toolset config that works on multiple types of hosts without bundles, whereas this approach you're suggesting will require separate buildtime-specific toolset configs.

I think this is worth consolidating in the toolset spec now, as the location of the build tools for different build-time triples is a single line that shouldn't require a whole new destination that is identical except for that, but up to you.

Reply:
Ok, I think I see what you mean. Would a toolset at the level of artifacts reusable across multiple variants help this in case? I initially had it in one of the previous commits, but felt that was too many levels to keep in mind when maintaining toolsets.

Reply:
Unsure what you're proposing, but my suggestion is to add an optional dictionary to the toolset spec, that maps each of the `supportedTriples` to their own subdirectory of `toolsetRootPath`. That way SPM knows where all the build tools are and a single destination installed on linux/Windiws can support multiple build-time triples, just like on Darwin.

Reply:
Have you thought about how you want to handle this?